src/Rings/MPolyMap/flattenings.jl

```diff
@@ -244,7 +387,12 @@ end
 
 ### Getters 
 function (phi::RingFlattening)(x::RingElem)
-  return phi.orig_to_flat(x)
+  if !haskey(flat_counterparts(phi), x)
+    result = phi.orig_to_flat(x)
+    flat_counterparts(phi)[x] = result
+    return result
+  end
+ return flat_counterparts(phi)[x]::elem_type(codomain(phi))
```

Comment:
I'm not 100% sure that doing this is really performant. And maybe we should also find a way to turn it off, e.g. using keyword arguments. Apart from that: @fingolfin , do you think this `flat_counterpart` architecture that I set up here is a safe and correct application of the `WeakKeyIdDict`s? I will link again to this below.